### PR TITLE
txn: add logs for assertion failure

### LINKED ISSUE #12304: There should be logs when assertion fails
When an assertion fails there should be logs to make it clear what happened.

### PR DESCRIPTION
**Pull request by ekexium**
Signed-off-by: ekexium <ekexium@gmail.com>

### What is changed and how it works?

Issue Number: Close #12304

What's Changed:

```commit-message
Add logs for assertion failure
```

The log would look like:
```
[2022/03/31 11:14:43.197 +08:00] [ERROR] [prewrite.rs:520] ["assertion failure"] [mutation="PrewriteMutation { key: 6F6B320000000000FA, value: Some([118, 50]), mutation_type: Put, secondary_keys: None, min_commit_ts: TimeStamp(0), is_pessimistic_lock: true, lock_type: Some(Put), lock_ttl: 0, should_not_exist: false, should_not_write: false, assertion: NotExist, txn_props: TransactionProperties { start_ts: TimeStamp(40), kind: Pessimistic(TimeStamp(41)), commit_kind: TwoPc, primary: [111, 107, 50], txn_size: 0, lock_ttl: 0, min_commit_ts: TimeStamp(0), need_old_value: false, is_retry_request: false, assertion_level: Strict } }"] [commit_ts=25] [write="Some(Write { write_type: Put, start_ts: TimeStamp(20), short_value: \"7632\", has_overlapped_rollback: false, gc_fence: None })"] [assertion=NotExist]
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Manual test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

File: src/storage/txn/actions/prewrite.rs [hunk 1]
```diff
@@ -501,12 +501,15 @@ impl<'a> PrewriteMutation<'a> {
 
         match (self.assertion, write) {
             (Assertion::Exist, None) => {
+                info!("assertion failed: exist, previous write record not found";);
```
Comment (cfzjywxk) [inline src/storage/txn/actions/prewrite.rs:504]: Better to print more information about the current transction, prewriting key/value pair and important flags such as `is_pessimisitic` key etc.

File: src/storage/txn/actions/prewrite.rs [hunk 2]
```diff
@@ -499,17 +500,26 @@ impl<'a> PrewriteMutation<'a> {
...
             }
-            _ => (),
+            _ => Ok(()),
+        };
+        if res.is_err() {
+            let (write, commit_ts) = write
+                .clone()
+                .map(|(w, ts)| (Some(w), Some(ts)))
+                .unwrap_or((None, None));
+            error!("assertion failure"; "assertion" => ?self.assertion, "write" => ?write,
```
Comment (cfzjywxk) [inline src/storage/txn/actions/prewrite.rs:520]: We may need to cover it by adding a test case or manually, and put the log output information in the pr description.

Comment (ekexium) [reply src/storage/txn/actions/prewrite.rs:520]: Yes. I've done that.

File: src/storage/txn/actions/prewrite.rs [hunk 3]
```diff
@@ -499,17 +500,26 @@ impl<'a> PrewriteMutation<'a> {
...
-            _ => (),
+            _ => Ok(()),
+        };
+        if res.is_err() {
+            let (write, commit_ts) = write
+                .clone()
+                .map(|(w, ts)| (Some(w), Some(ts)))
+                .unwrap_or((None, None));
+            error!("assertion failure"; "assertion" => ?self.assertion, "write" => ?write, 
+            "commit_ts" => commit_ts, "mutation" => ?self);
```
Comment (MyonKeminta) [inline src/storage/txn/actions/prewrite.rs:521]: Consider also print the error itself.

File: src/storage/txn/actions/prewrite.rs [hunk 4]
```diff
@@ -499,17 +500,26 @@ impl<'a> PrewriteMutation<'a> {
...
+                self.assertion_failed_error(w.start_ts, *commit_ts)
             }
-            _ => (),
+            _ => Ok(()),
+        };
+        if res.is_err() {
+            let (write, commit_ts) = write
+                .clone()
+                .map(|(w, ts)| (Some(w), Some(ts)))
+                .unwrap_or((None, None));
```
Comment (MyonKeminta) [inline src/storage/txn/actions/prewrite.rs:519]: Must it be cloned? Couldn't it be `write.as_ref().map ....` ?

File: src/storage/txn/actions/prewrite.rs (hunk 3 above)
Comment (ekexium) [reply src/storage/txn/actions/prewrite.rs:521]: All the information in the error is included. I think it is redundant to print the error.

**Review by cfzjywxk — APPROVED** [review]

**Comment by ekexium** [comment]
@MyonKeminta PTAL

**Review by MyonKeminta — APPROVED** [review]

**Pull request merged**